Add the ability to run squiggle files straight

noahmorrison opened the pull request:
Calling squiggle with only one argument will eval the compiled
code instead of writting it to a file.

This also allows squiggle to be used from a shebang

> example.sql
> 
> ``` shell
> #!/usr/bin/env squiggle
> global.console.log("Hello, World!")
> ```

```
$ chmod +x example.sql
$ ./example.sql
Hello, World!
```

noahmorrison committed:
Add the ability to run squiggle files straight

Calling squiggle with only one argument will eval the compiled
code instead of writting it to a file.

This also allows squiggle to be used from a shebang

```example.sql
#!/usr/bin/env squiggle
global.console.log("Hello, World!")
```

$ chmod +x example.sql
$ ./example.sql
Hello, World!

wavebeem commented:
I'm using the extension `.sqg` to avoid confusion with SQL.

I like this idea, but you can't import other Squiggle files when using this. I think this feature would need to use [node-hook](https://github.com/bahmutov/node-hook) otherwise it'd be confusing.

wavebeem commented:
Also, please look up how to quiet the warning about "eval" from jshint on that line so that it passes TravisCI.

noahmorrison committed:
Use vm instead of eval

eval is evil, vm is viably malicious.

wavebeem commented on src/main.js:
```diff
@@ -18,7 +18,7 @@ var nomnom = require("nomnom")
         position: 1,
         metavar: "FILE",
-        help: "Write JavaScript to this file",
+        help: "Write JavaScript to this file, if omitted then run the input",
```
This is kinda confusing. I think if Squiggle can run code instead of just compiling it, this should be moved into like a `-o FILE` option which can be omitted.

wavebeem commented on src/main.js:
```diff
@@ -92,6 +95,8 @@ function compileTo(src, dest) {
 if (argv.interactive) {
     repl.start();
+} else if (argv._.length === 1) {
+    compileTo(argv._[0], undefined);
```
It's kinda weird to use this function to either compile or eval. How about splitting this function into one that returns the compiled code and call it `compile`, then have like `compileAndEval` and `compileToDest` use it in this if/else area?

noahmorrison replied:
There's already a function called `compile` in compile.js, any other naming suggestions?

wavebeem replied:
Just call it `compileHelper` or something.

On Tue, Nov 17, 2015 at 4:09 PM Noah Morrison notifications@github.com
wrote:

> In src/main.js
> https://github.com/wavebeem/squiggle/pull/113#discussion_r45142271:
> 
> > @@ -92,6 +95,8 @@ function compileTo(src, dest) {
> > 
> >  if (argv.interactive) {
> >      repl.start();
> > +} else if (argv._.length === 1) {
> > -    compileTo(argv._[0], undefined);
> 
> There's already a function called compile in compile.js, any other naming
> suggestions?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/wavebeem/squiggle/pull/113/files#r45142271.

wavebeem commented on src/main.js:
```diff
@@ -56,7 +56,6 @@ function die(message) {
 }
 
 function compileTo(src, dest) {
-    var jsOut = dest;
```
Nice catch on my useless var here.

noahmorrison commented:
I got travis to be quiet about eval by using vm instead, I'll look into the other things right now

wavebeem commented:
Hmm, I forgot about the VM module. I wonder if you can use `node-hook` with that? What I had meant was simply putting a jshint magic comment to ignore the use of eval there, but this may be better if it works out in the end.

noahmorrison committed:
Make cli options less confusing

If you want to output to a file, then use the -o|--output option
instead of using a second, optional, parameter.

noahmorrison committed:
Use node-hook to fix requiring squiggle files

When a file is run without compilation it would not be able to
require a file. node-hook allows us to setup custom compilers for a
file extension.

Also setup correct __filename and __dirname globals in the sandbox
that runs squiggle code.

noahmorrison commented:
This was a lot harder then I thought it would be, but I believe it works now as I had originally intended. I used node-hook like you suggested to handle preprocessing required squiggle files. And I also setup the context for the vm.Script. I don't think there should be any difference between running a squiggle file directly and compiling it. If there is then it's a bug.

It's important to note that vm.Script::runInThisContext leaks functions into the code, so we need to use runInContext and set up the globals ourselves.

Play around with it, see if you spot anything I missed :smile:

noahmorrison commented:
Oh, another thing, do you know why my `npm link` command linked it to `/usr/local/bin/squiggle-lang` instead of `/usr/local/bin/squiggle`? Just something that's been bugging me :stuck_out_tongue_closed_eyes: 

EDIT: Ahh, [this is the line](https://github.com/wavebeem/squiggle/blob/master/package.json#L2). I was just curious as to why it deviated from the [readme](https://github.com/wavebeem/squiggle/blob/master/README.md#contributing).

noahmorrison commented:
Ahh, [this is the line](https://github.com/wavebeem/squiggle/blob/master/package.json#L2). I was just curious as to why it deviated from the [readme](https://github.com/wavebeem/squiggle/blob/master/README.md#contributing).

wavebeem commented:
Oh, nice catch on the wrong bin. The package used to be named `squiggle` not `squiggle-lang`, so I never fixed that somehow.

wavebeem commented on src/main.js:
```diff
@@ -5,21 +5,25 @@ var UTF8 = "utf-8";
         help: "Write JavaScript to this file",
-        type: "string"
+        type: "string",
+        flag: true
```
This shouldn't be a `flag` option. It should be be a regular parameter with a value `-o FILE`. I'm fine if this breaks existing calls to `squiggle main.sqg out.js`.

wavebeem commented on src/main.js:
```diff
@@ -90,8 +93,39 @@ function compileTo(src, dest) {
+} else if (argv._.length === 1) {
+    compileAndRun(argv._[0]);
 } else if (argv._.length === 2) {
     compileTo(argv._[0], argv._[1]);
```
Move the `compileTo` above the previous `else if` and check for `argv.output` and use that value.

wavebeem commented on src/main.js:
```diff
@@ -90,8 +93,39 @@ function compileTo(src, dest) {
+        }
+    };
+    root.__filename = path.normalize(path.join(process.cwd(), src));
+    root.__dirname = path.dirname(root.__filename) + "/";
```
I don't think this needs the `/` at the end. If I do `node -p "__dirname"` it simply prints out `.`.

wavebeem commented on src/main.js:
```diff
@@ -90,8 +93,39 @@ function compileTo(src, dest) {
+    var js = compileSource(src);
+    var script = new vm.Script(js);
+    var root = global;
+    root.require = function(id) {
```
I think it should be sufficient here to check for whether `id` starts with `.`, meaning it's a local import instead of one managed by npm. Also, paths should be relative to `src`, so this needs to grab the dirname of `src`.

wavebeem commented:
I did some more research into this, and it generally seems frowned upon in the Node community to let `require` load anything besides `.js`, `.json`, and `.node` files. Sadly, I think maybe this feature isn't meant to exist for these reasons.
- https://github.com/nodejs/node-v0.x-archive/issues/5430
- https://nodejs.org/api/globals.html#globals_require_extensions

wavebeem commented:
I think the ideal implementation here is something like TypeScript's `--outDir` command where you can do something like:

```
tsc src/main.ts --outDir build
node build/main.js
```

Or maybe just something like:

```
squiggle -i src -o build
node build/main.js
```

That could just copy src to build recursively, then walk the tree and convert all the .sqg to .js. Probably best if Squiggle stays out of the "running JS" and "loading modules" businesses. Those are both tricky, and not something we're really meant to be mucking with.

wavebeem commented:
I'm starting to think this isn't a great idea, per my comment above. If you wanted to work on the `squiggle -i src -o build` thing instead I think that would be cool. I'm gonna make an issue around the idea and tag you in it in case you're interested.
